<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting Started • romicidx</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="omicidx_files/libs/quarto-html/popper.min.js"></script><script src="omicidx_files/libs/quarto-html/tippy.umd.min.js"></script><link href="omicidx_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="omicidx_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started">
<!-- Google tag (gtag.js) --><script async src="https://www.googletagmanager.com/gtag/js?id=G-EWN5JN9XKZ"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-EWN5JN9XKZ');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">romicidx</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/omicidx.html">Getting Started</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/seandavi/romicidx/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Getting Started</h1>




      <small class="dont-index">Source: <a href="https://github.com/seandavi/romicidx/blob/main/vignettes/omicidx.qmd" class="external-link"><code>vignettes/omicidx.qmd</code></a></small>
      <div class="d-none name"><code></code></div>
    </div>






    <div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://seandavi.github.io/romicidx/">omicidx</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: DBI</span></span>
<span><span class="co">#&gt; Loading required package: duckdb</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; Loading required package: dbplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dbplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     ident, sql</span></span>
<span><span class="co">#&gt; Loading required package: R6</span></span></code></pre></div>
</div>
<p>The <code>omicidx</code> package provides an interface to the OmicIDX DuckDB database, which contains a wealth of omics data. This vignette demonstrates how to establish a connection to the database and perform basic queries.</p>
<section class="level2"><h2 id="establishing-a-connection">Establishing a Connection<a class="anchor" aria-label="anchor" href="#establishing-a-connection"></a>
</h2>
<p>To connect to the OmicIDX DuckDB database, use the <code><a href="../reference/omicidx_duckdb_connection.html">omicidx_duckdb_connection()</a></code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">con</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/omicidx_duckdb_connection.html">omicidx_duckdb_connection</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">con</span></span>
<span><span class="co">#&gt; &lt;duckdb_connection 7ba50 driver=&lt;duckdb_driver dbdir=':memory:' read_only=FALSE bigint=numeric&gt;&gt;</span></span></code></pre></div>
</div>
<p>Note that this function sets up a read-only connection to the remote DuckDB database hosted remotely. The database doesn’t actually contain any data. Instead it contains links to data stored as parquet files in cloud storage.</p>
</section><section class="level2"><h2 id="listing-tables">Listing Tables<a class="anchor" aria-label="anchor" href="#listing-tables"></a>
</h2>
<p>Once connected, you can list the available tables in the database:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tables</span> <span class="op">&lt;-</span> <span class="fu">DBI</span><span class="fu">::</span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbListTables.html" class="external-link">dbListTables</a></span><span class="op">(</span><span class="va">con</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">tables</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "_catalog"             "ncbi_biosamples"      "sra_experiments"     </span></span>
<span><span class="co">#&gt;  [4] "sra_metadata"         "sra_runs"             "sra_samples"         </span></span>
<span><span class="co">#&gt;  [7] "sra_studies"          "stg_geo_platforms"    "stg_geo_samples"     </span></span>
<span><span class="co">#&gt; [10] "stg_geo_series"       "stg_ncbi_bioprojects" "stg_ncbi_biosamples" </span></span>
<span><span class="co">#&gt; [13] "stg_sra_accessions"   "stg_sra_experiments"  "stg_sra_runs"        </span></span>
<span><span class="co">#&gt; [16] "stg_sra_samples"      "stg_sra_studies"</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="querying-data">Querying Data<a class="anchor" aria-label="anchor" href="#querying-data"></a>
</h2>
<p>Each table in the database can be queried using standard SQL. For example, to retrieve the first 10 rows from the <code>sra_runs</code> table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">DBI</span><span class="fu">::</span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbGetQuery.html" class="external-link">dbGetQuery</a></span><span class="op">(</span><span class="va">con</span>, <span class="st">"SELECT * FROM staging.stg_sra_runs LIMIT 10"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt;     accession experiment_accession title run_center run_date center_name</span></span>
<span><span class="co">#&gt; 1  SRR4652455           SRX2310319  &lt;NA&gt;       &lt;NA&gt;     &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2  SRR4661781           SRX2310320  &lt;NA&gt;       &lt;NA&gt;     &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 3  SRR4661782           SRX2310321  &lt;NA&gt;       &lt;NA&gt;     &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 4  SRR4661783           SRX2310322  &lt;NA&gt;       &lt;NA&gt;     &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 5  SRR4661784           SRX2310323  &lt;NA&gt;       &lt;NA&gt;     &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 6  SRR4661785           SRX2310324  &lt;NA&gt;       &lt;NA&gt;     &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 7  SRR4661786           SRX2310325  &lt;NA&gt;       &lt;NA&gt;     &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 8  SRR4661787           SRX2310326  &lt;NA&gt;       &lt;NA&gt;     &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 9  SRR4661788           SRX2310327  &lt;NA&gt;       &lt;NA&gt;     &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 10 SRR4661789           SRX2310328  &lt;NA&gt;       &lt;NA&gt;     &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt;    broker_name                                        alias  GEO</span></span>
<span><span class="co">#&gt; 1         &lt;NA&gt; P0028_MT_I2722_TTCATACG_L001_R1_001.fastq.gz &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2         &lt;NA&gt;             FSL_E2-0214_R1.trimmedP.fastq.gz &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 3         &lt;NA&gt;             FSL_K6-1142_R2.trimmedP.fastq.gz &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 4         &lt;NA&gt;             FSL_K6-1030_R2.trimmedP.fastq.gz &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 5         &lt;NA&gt;             FSL_H8-0481_R1.trimmedP.fastq.gz &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 6         &lt;NA&gt;             FSL_H7-0909_R1.trimmedP.fastq.gz &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 7         &lt;NA&gt;             FSL_H7-0676_R1.trimmedP.fastq.gz &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 8         &lt;NA&gt;             FSL_F4-0079_R2.trimmedP.fastq.gz &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 9         &lt;NA&gt;             FSL_H8-0063_R1.trimmedP.fastq.gz &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 10        &lt;NA&gt;             FSL_H8-0534_R2.trimmedP.fastq.gz &lt;NA&gt;</span></span>
<span><span class="co">#&gt;                                                     identifiers attributes</span></span>
<span><span class="co">#&gt; 1  P0028_MT_I2722_TTCATACG_L001_R1_001.fastq.gz, SUB2043882, NA       NULL</span></span>
<span><span class="co">#&gt; 2              FSL_E2-0214_R1.trimmedP.fastq.gz, SUB2043904, NA       NULL</span></span>
<span><span class="co">#&gt; 3              FSL_K6-1142_R2.trimmedP.fastq.gz, SUB2043904, NA       NULL</span></span>
<span><span class="co">#&gt; 4              FSL_K6-1030_R2.trimmedP.fastq.gz, SUB2043904, NA       NULL</span></span>
<span><span class="co">#&gt; 5              FSL_H8-0481_R1.trimmedP.fastq.gz, SUB2043904, NA       NULL</span></span>
<span><span class="co">#&gt; 6              FSL_H7-0909_R1.trimmedP.fastq.gz, SUB2043904, NA       NULL</span></span>
<span><span class="co">#&gt; 7              FSL_H7-0676_R1.trimmedP.fastq.gz, SUB2043904, NA       NULL</span></span>
<span><span class="co">#&gt; 8              FSL_F4-0079_R2.trimmedP.fastq.gz, SUB2043904, NA       NULL</span></span>
<span><span class="co">#&gt; 9              FSL_H8-0063_R1.trimmedP.fastq.gz, SUB2043904, NA       NULL</span></span>
<span><span class="co">#&gt; 10             FSL_H8-0534_R2.trimmedP.fastq.gz, SUB2043904, NA       NULL</span></span>
<span><span class="co">#&gt;    qualities has_complete_run_info          _loaded_at</span></span>
<span><span class="co">#&gt; 1                            FALSE 2025-10-30 23:06:34</span></span>
<span><span class="co">#&gt; 2                            FALSE 2025-10-30 23:06:34</span></span>
<span><span class="co">#&gt; 3                            FALSE 2025-10-30 23:06:34</span></span>
<span><span class="co">#&gt; 4                            FALSE 2025-10-30 23:06:34</span></span>
<span><span class="co">#&gt; 5                            FALSE 2025-10-30 23:06:34</span></span>
<span><span class="co">#&gt; 6                            FALSE 2025-10-30 23:06:34</span></span>
<span><span class="co">#&gt; 7                            FALSE 2025-10-30 23:06:34</span></span>
<span><span class="co">#&gt; 8                            FALSE 2025-10-30 23:06:34</span></span>
<span><span class="co">#&gt; 9                            FALSE 2025-10-30 23:06:34</span></span>
<span><span class="co">#&gt; 10                           FALSE 2025-10-30 23:06:34</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="usage-with-dplyr">Usage with dplyr<a class="anchor" aria-label="anchor" href="#usage-with-dplyr"></a>
</h2>
<p>The connection can also be used with the <code>dplyr</code> package for more convenient data manipulation. For example, to get the first 10 entries from the <code>sra_studies</code> table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sra_studies_tbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/tbl.html" class="external-link">tbl</a></span><span class="op">(</span><span class="va">con</span>, <span class="st">"staging.stg_sra_studies"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">sra_studies_tbl</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/compute.html" class="external-link">collect</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 10 × 17</span></span>
<span><span class="co">#&gt;    accession study_accession title   description abstract study_type center_name</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;chr&gt;           &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;      </span></span>
<span><span class="co">#&gt;  1 SRP009395 SRP009395       Subtro… &lt;NA&gt;        The sco… Metagenom… Biodiversi…</span></span>
<span><span class="co">#&gt;  2 SRP009396 SRP009396       Identi… &lt;NA&gt;        The aim… Other      GEO        </span></span>
<span><span class="co">#&gt;  3 SRP009397 SRP009397       Entamo… &lt;NA&gt;        Parasit… Whole Gen… JCVI       </span></span>
<span><span class="co">#&gt;  4 SRP009399 SRP009399       Genome… &lt;NA&gt;        Backgro… Whole Gen… Umass Medi…</span></span>
<span><span class="co">#&gt;  5 SRP009401 SRP009401       Transc… &lt;NA&gt;        Sacchar… Transcrip… GEO        </span></span>
<span><span class="co">#&gt;  6 ERP001015 ERP001015       Antarc… Pyrosequen… The ext… Metagenom… ICTAR      </span></span>
<span><span class="co">#&gt;  7 SRP009402 SRP009402       Charac… &lt;NA&gt;        The pre… Transcrip… Graduate U…</span></span>
<span><span class="co">#&gt;  8 SRP009404 SRP009404       Transc… &lt;NA&gt;        To unde… Transcrip… Shanghai O…</span></span>
<span><span class="co">#&gt;  9 SRP009407 SRP009407       Cheese… &lt;NA&gt;        Grana T… Metagenom… IASMA rese…</span></span>
<span><span class="co">#&gt; 10 SRP009408 SRP009408       MicroR… &lt;NA&gt;        Backgro… Transcrip… GEO        </span></span>
<span><span class="co">#&gt; # ℹ 10 more variables: broker_name &lt;chr&gt;, alias &lt;chr&gt;, BioProject &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   GEO &lt;chr&gt;, pubmed_ids &lt;list&gt;, has_complete_metadata &lt;lgl&gt;,</span></span>
<span><span class="co">#&gt; #   attributes &lt;list&gt;, identifiers &lt;list&gt;, xrefs &lt;list&gt;, `_loaded_at` &lt;dttm&gt;</span></span></code></pre></div>
</div>
<p>We can do more complex queries as well, such as grouping, counting, and arranging data. For example, to count the number of studies by study type:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">studies_by_type</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">sra_studies_tbl</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">study_type</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/compute.html" class="external-link">collect</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">studies_by_type</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 12 × 2</span></span>
<span><span class="co">#&gt; # Groups:   study_type [12]</span></span>
<span><span class="co">#&gt;    study_type                      n</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                       &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 "Other"                    364910</span></span>
<span><span class="co">#&gt;  2 "Whole Genome Sequencing"  119980</span></span>
<span><span class="co">#&gt;  3 "Transcriptome Analysis"   112895</span></span>
<span><span class="co">#&gt;  4 "Metagenomics"              64548</span></span>
<span><span class="co">#&gt;  5 "Population Genomics"         835</span></span>
<span><span class="co">#&gt;  6 "Epigenetics"                 702</span></span>
<span><span class="co">#&gt;  7 "Cancer Genomics"             429</span></span>
<span><span class="co">#&gt;  8 "Exome Sequencing"            372</span></span>
<span><span class="co">#&gt;  9 "Pooled Clone Sequencing"      31</span></span>
<span><span class="co">#&gt; 10 "Synthetic Genomics"           12</span></span>
<span><span class="co">#&gt; 11 "Transcriptome Sequencing"      2</span></span>
<span><span class="co">#&gt; 12 "Whole Genome Sequencing "      1</span></span></code></pre></div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config);
    }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sean Davis.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
